src/RequestHandler/outofprocess/forwardinghandler.cpp

```diff
@@ -536,103 +568,133 @@ REQUEST_NOTIFICATION_STATUS
+                    NULL);
+            }
+            else
+            {
+                LoadString(g_hAspNetCoreModule,
+                    IDS_SERVER_ERROR,
+                    strDescription.QueryStr(),
+                    strDescription.QuerySizeCCH());
+            }
+            (VOID)strDescription.SyncWithBuffer();
 
-        (VOID)strDescription.SyncWithBuffer();
-        if (strDescription.QueryCCH() != 0)
-        {
-            pResponse->SetErrorDescription(
-                strDescription.QueryStr(),
-                strDescription.QueryCCH(),
-                FALSE);
+            if (strDescription.QueryCCH() != 0)
+            {
+                pResponse->SetErrorDescription(
+                    strDescription.QueryStr(),
+                    strDescription.QueryCCH(),
+                    FALSE);
+            }
+
+            if (hr == HRESULT_FROM_WIN32(ERROR_WINHTTP_INVALID_SERVER_RESPONSE))
+            {
+                if (!m_fServerResetConn)
+                {
+                    RemoveRequest();
+                    pResponse->ResetConnection();
+                    m_fServerResetConn = TRUE;
+                }
+            }
         }
     }
 
-    //
-    // Finish the request on failure.
-    // Let IIS pipeline continue only after receiving handle close callback
-    // from WinHttp. This ensures no more callback from WinHttp
-    //
-    if (m_hRequest != NULL)
+    if (m_pWebSocket != NULL && !m_fWebSocketHandleInClose)
     {
-        if (WinHttpCloseHandle(m_hRequest))
-        {
-            m_hRequest = NULL;
-        }
-        else
-        {
-            // Failed to close the handle 
-            // which should never happen as we registered a callback with WinHttp
-            // For this unexpected failure, let conitnue IIS pipeline
-           /* retVal = RQ_NOTIFICATION_FINISH_REQUEST;
-            DebugBreak();*/
-        }
+        m_fWebSocketHandleInClose = TRUE;
+        m_pWebSocket->TerminateRequest();
+    }
+
+    if (m_hRequest != NULL && !m_fHttpHandleInClose)
+    {
+        m_fHttpHandleInClose = TRUE;
+        WinHttpCloseHandle(m_hRequest);
+        m_hRequest = NULL;
     }
-    retVal = RQ_NOTIFICATION_PENDING;
 
 Finished:
+
     if (fLocked)
     {
         DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == this);
         TlsSetValue(g_dwTlsIndex, NULL);
-        ReleaseSRWLockShared(&m_RequestLock);
+        ReleaseSRWLockExclusive(&m_RequestLock);
         DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == NULL);
     }
 
+    if (retVal != RQ_NOTIFICATION_PENDING)
+    {
+
+        DBG_ASSERT(m_dwHandlers == 0);
+        RemoveRequest();
+
+        // This is just a safety guard to prevent from returning non pending status no more once
+        // which should never happen
+        if (!m_fDoneAsyncCompletion)
+        {
+            m_fDoneAsyncCompletion = TRUE;
+        }
+        else
+        {
+            retVal = RQ_NOTIFICATION_PENDING;
```

Comment:
I don't understand why we are returning PENDING instead of CONTINUE here.

Reply:
this logic is that current thread hit an error (ideally should return non-pending status to IIS). However, another thread already returned (as m_fDoneAsyncCompletion was true). Current thread cannot return non-pending anymore otherwise IIS pipe will get messed as the later may already move to another module